### ISSUE TITLE
Has many through doesn't reify children

### ISSUE THREAD

**Issue by tlynam** [airblade/paper_trail#633]
Here's a failing test showing when has_many through's source association is destroyed, it can no longer be accessed via `reify`.

```diff
diff --git a/test/unit/associations_test.rb b/test/unit/associations_test.rb
index ef70ed6..2cc05c6 100644
--- a/test/unit/associations_test.rb
+++ b/test/unit/associations_test.rb
@@ -633,7 +633,20 @@ class AssociationsTest < ActiveSupport::TestCase
             end
           end

-          context "destroy a section" do
+          context "the version before a section is destroyed" do
+            should "have the section and paragraph" do
+              @chapter.update_attributes(:name => 'chapter 6')
+              Timecop.travel 1.second.since
+              @section.destroy
+              assert_equal 4, @chapter.versions.size
+              chapter_v3 = @chapter.versions[3].reify(:has_many => true)
+              assert_equal 1, chapter_v3.sections.size
+              assert_equal 1, chapter_v3.paragraphs.size
+            end
+          end
+
+          context "the version after a section is destroyed" do
             should "not have any sections or paragraphs" do
               @section.destroy
               Timecop.travel 1.second.since
```

**Comment by jaredbeck** [comment]
Thanks for the failing test. I think we want it to pass, as written, so I'm marking this issue a confirmed bug.

I've spent a few hours playing around with `reify` and I think this may be a symptom of a larger issue. It seems that, when restoring the HMT association `Chapter#paragraphs`, the "through" part of the association (sections) is reified, but the final "target" part of the association (paragraphs) is not.

I've written a branch that makes the test pass, [fix_issue_633](https://github.com/airblade/paper_trail/tree/fix_issue_633), but reification is such a complicated process that I don't even have the confidence to make a PR out of it. Please take a look at the branch and tell me if you think I'm on the right track.

CC: @batter

**Comment by batter** [comment]
What is the proposal for handling scenarios like this? Have a `dependent: :destroy` like cascade on the associated verisons when a joiner model is destroyed in a relationship like this?

**Comment by jaredbeck** [comment]
> What is the proposal for handling scenarios like this? 

I'm not sure I understand the question.  Did you see my branch? https://github.com/airblade/paper_trail/tree/fix_issue_633

> Have a dependent: :destroy like cascade on the associated versions when a joiner model is destroyed in a relationship like this?

The Chapter/Section/Paragraph models already have `:dependent => :destroy`, but actually cascade deletion is not the problem.  The problem is refiying a version from before a cascade delete.  See Todd's failing test.

**Comment by batter** [comment]
@jaredbeck - Apologies, for some reason when I first glanced I thought that the through model being destroyed resulted in a scenario where the model is unable to be reified.  Disregard my question from the previous post.

> I'm not sure I understand the question. Did you see my branch?
I just inspected the branch more closely. It seems pretty spot on as far as I can tell, but as you mentioned reification of nested associations is pretty complex, and I don't know that I'm that much more well versed than you in this stuff. That being said, it does make the test pass and the logic seems sound for temporarily reifying / reaching down through a cascading tree and appears to work.
